File: Depth-First Search/Memoization/LET_no1066_Campus Bikes II.py
class Solution:
    def assignBikes(self, workers: List[List[int]], bikes: List[List[int]]) -> int:
        memo = {}
        used = [False] * len(bikes)
        return self.dfs(0, workers, bikes, used, memo)

    def dfs(self, i, workers, bikes, used, memo):
        if i == len(workers):
            return 0

        if (i, tuple(used)) not in memo:
            cur = float('inf')
            for j in range(len(bikes)):
                if not used[j]:
                    used[j] = True
                    cur = min(cur, self.get_dist(workers[i], bikes[j]) + self.dfs(i + 1, workers, bikes, used, memo))
                    used[j] = False
            memo[(i, tuple(used))] = cur
        
        return memo[(i, tuple(used))]

    def get_dist(self, a, b):
        return abs(a[0] - b[0]) + abs(a[1] - b[1])
        

File: Depth-First Search/Memoization/LET_no1066_Campus Bikes II.cpp
class Solution {
public:
    int assignBikes(vector<vector<int>>& workers, vector<vector<int>>& bikes) {
        vector<int> memo(1 << bikes.size(), INT_MIN);
        return dfs(0, workers, bikes, 0, memo);
    }

    int dfs(int i, vector<vector<int>>& workers, vector<vector<int>>& bikes, int used_bikes, vector<int>& memo) {
        if (i == workers.size())
            return 0;

        if (memo[used_bikes] == INT_MIN) {
            int cur = INT_MAX;
            for (int j = 0; j < bikes.size(); j++) {
                if (used_bikes & (1 << j))
                    continue;
                cur = min(cur, getDist(workers[i], bikes[j]) + dfs(i + 1, workers, bikes, used_bikes | (1 << j), memo));
            }
            memo[used_bikes] = cur;
        }

        return memo[used_bikes];
    }

    int getDist(vector<int>& a, vector<int>& b) {
        return abs(a[0] - b[0]) + abs(a[1] - b[1]);
    }
};